dlio/h5bench_dlio.c

```diff
@@ -0,0 +1,805 @@
+        }
+        for (uint32_t j = 0; j < config.BATCH_SIZE; j++) {
+            uint32_t file_num = indices[offset + i * config.BATCH_SIZE + j] / config.NUM_SAMPLES_PER_FILE + 1;
+            uint32_t sample_num = indices[offset + i * config.BATCH_SIZE + j] % config.NUM_SAMPLES_PER_FILE;
+            char     file_path[256];
+            snprintf(file_path, sizeof(file_path), "%s/%s/%s_%u_of_%u.h5", config.DATA_FOLDER,
+                     config.TRAIN_DATA_FOLDER, config.FILE_PREFIX, file_num, config.NUM_FILES_TRAIN);
+
+            uint64_t metadata_time = 0, read_time = 0;
+            read_sample(file_path, sample_num, &metadata_time, &read_time);
+
+            *local_metadata_time_out += metadata_time;
+            *local_read_time_out += read_time;
+        }
+
+        batch_loaded_train(epoch, t0);
+
+        uint64_t t = compute(config.COMPUTATION_TIME, config.COMPUTATION_TIME_STDEV);
+        batch_processed_train(epoch, t, t0);
+        if ((MY_RANK < config.TOTAL_TRAINING_STEPS % NUM_RANKS) &&
+            (i + 1 == config.TOTAL_TRAINING_STEPS_PER_RANK)) {
+            MPI_Barrier(rest_training_steps_comm);
+        }
+        else {
+            MPI_Barrier(MPI_COMM_WORLD);
+        }
+
+        t0 = get_time_usec_return_uint64();
+    }
+
+    //    TODO: drop_data = True
+    //    for (uint32_t iteration = MY_RANK; iteration < total_samples - NUM_RANKS * samples_per_rank;
+    //    iteration += NUM_RANKS) {
+    //        uint32_t i = NUM_RANKS * samples_per_rank + iteration;
+    //        uint32_t file_num = indices[i] / config.NUM_SAMPLES_PER_FILE + 1;
+    //        uint32_t sample_num = indices[i] % config.NUM_SAMPLES_PER_FILE;
+    //        char file_path[256];
+    //        snprintf(file_path, sizeof(file_path), "%s/%s/%s_%u_of_%u.h5", config.DATA_FOLDER,
+    //        config.TRAIN_DATA_FOLDER, config.FILE_PREFIX, file_num, config.NUM_FILES_TRAIN);
+    //
+    //        uint64_t metadata_time = 0, read_time = 0;
+    //        read_sample(file_path, sample_num, &metadata_time, &read_time);
+    //        read_counter++;
+    //
+    //        *local_metadata_time_out += metadata_time;
+    //        *local_read_time_out += read_time;
+    //
+    //        if (read_counter % config.BATCH_SIZE == 0){
+    //            batch_loaded_train(epoch, t0);
+    //
+    //            uint64_t t = compute(config.COMPUTATION_TIME, config.COMPUTATION_TIME_STDEV);
+    //            batch_processed_train(epoch, t, t0);
+    //
+    //            read_counter = 0;
+    //            t0 = get_time_usec_return_uint64();
+    //        }
+    //    }
+    //
+    //    if (read_counter != 0) {
+    //        batch_loaded_train(epoch, t0);
+    //
+    //        uint64_t t = compute(config.COMPUTATION_TIME, config.COMPUTATION_TIME_STDEV);
+    //        batch_processed_train(epoch, t, t0);
+    //    }
+}
+
+void
+train_using_workers(uint32_t epoch, uint64_t *local_metadata_time_out, uint64_t *local_read_time_out)
```

Comment:
Could we also add some short in-code documentation about what each function is doing?